src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,854 @@
+    > Gao, Bin and Son, Nguyen Thanh and Absil, P-A and Stykel, Tatjana:
+    > Riemannian optimization on the symplectic Stiefel manifold,
+    > SIAM Journal on Optimization 31(2), pp. 1546-1575, 2021.
+    > doi [10.1137/20M1348522](https://doi.org/10.1137/20M1348522)
+"""
+function project_riemannian_normal!(::Symplectic, Y, p, X)
+    Q = SymplecticMatrix(p, X)
+
+    pT_QT_X = p' * Q' * X
+    skew_pT_QT_X = (1 / 2) .* (pT_QT_X .- pT_QT_X')
+
+    Y[:, :] = p * Q * skew_pT_QT_X
+    return Y
+end
+
+function Base.rand(M::Symplectic{n}, hamiltonian_norm::Number=1.0) where {n}
+    # Generate random matrices to construct a Hamiltonian matrix:
+    Ω = rand_hamiltonian(M; final_norm=hamiltonian_norm)
+    # Return 'cay(Ω)':
+    return (I - Ω) \ (I + Ω)
+end
+
+function Base.rand(::Symplectic{n}, p::AbstractMatrix) where {n}
+    # Generate random symmetric matrix:
+    S = rand(2n, 2n) .- 1 / 2
+    S .= (1 / 2) .* (S + S')
+    Q = SymplecticMatrix(p)
+    lmul!(Q, S)
+    return p * S
+end
+
+function rand_hamiltonian(::Symplectic{n}; final_norm=1) where {n}
+    A = randn(n, n)
+    B = randn(n, n)
+    C = randn(n, n)
+    B = (1 / 2) .* (B .+ B')
+    C = (1 / 2) .* (C .+ C')
+    Ω = [A B; C -A']
+    return final_norm * Ω / norm(Ω, 2)
+end
+
+@doc raw"""
+    retract(::Symplectic, p, X, ::CayleyRetraction)
+    retract!(::Symplectic, q, p, X, ::CayleyRetraction)
+
+Compute the Cayley retraction on ``p ∈ \operatorname{Sp}(2n, ℝ)`` in
+the direction of tangent vector ``X ∈ T_p\operatorname{Sp}(2n, ℝ)``,
+as defined in by Birtea et al in proposition 2 [^birtea2020optimization].
+
+Using the symplectic inverse of a matrix ``A \in ℝ^{2n \times 2n}``,
+``
+A^{+} := Q_{2n}^T A^T Q_{2n}
+``
+where
+````math
+Q_{2n} =
+\begin{bmatrix}
+0_n & I_n \\
+ -I_n & 0_n
+\end{bmatrix},
+````
+the retraction
+``\mathcal{R}\colon T\operatorname{Sp}(2n) \rightarrow \operatorname{Sp}(2n)``
+is defined pointwise as
+````math
+\begin{align*}
+\mathcal{R}_p(X) &= p \operatorname{cay}\left(\frac{1}{2}p^{+}X\right), \\
+                 &= p \operatorname{exp}_{1/1}(p^{+}X), \\
+                 &= p (2I - p^{+}X)^{-1}(2I + p^{+}X).
+\end{align*}
+````
+Here
+``\operatorname{exp}_{1/1}(z) = (2 - z)^{-1}(2 + z)``
+denotes the Padé (1, 1) approximation to ``\operatorname{exp}(z)``.
+
+[^birtea2020optimization]:
+    > Birtea, Petre and Ca{\c{s}}u, Ioan and Com{\u{a}}nescu, Dan:
+    > Optimization on the real symplectic group,
+    > Monatshefte f{\"u}r Mathematik, Springer
+    > doi [10.1007/s00605-020-01369-9](https://doi.org/10.1007/s00605-020-01369-9)
+"""
+function retract!(M::Symplectic, q, p, X, ::CayleyRetraction)
+    p_star_X = symplectic_inverse_times(M, p, X)
+
+    divisor = lu(2 * I - p_star_X)
+    q .= p * (divisor \ add_scaled_I!(p_star_X, 2.0))
+    return q
+end
+
+Base.show(io::IO, ::Symplectic{n,𝔽}) where {n,𝔽} = print(io, "Symplectic{$(2n), $(𝔽)}()")
+
+function symplectic_inverse_times(M::Symplectic{n}, p, q) where {n}
```

Comment:
This should be documented, I think

Reply:
I ported the documentation from the `SymplecticStiefel` case to this one now :+1: